### PR TITLE
Database service integration

### PR DESCRIPTION
**Pull request by Itxaka**
Missing:
- [x] ~~Spinner~~
- [x] ~~Tests~~
- [x] ~~Error handling~~ (out of scope)
- [x] ~~Running/complete states set~~
- [x] ~~Long running function to check database upgrade status after start~~
- [x] ~~Button disable on start~~

### CODE REVIEW

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 1]
```diff
@@ -69,9 +75,20 @@
...
                     },
                     // error
-                    function (/*errorResponse*/) {
+                    function (errorResponse) {
+                        vm.errors = getErrorFromResponse(errorResponse)
                     }
                 )
         }
+
+        function getErrorFromResponse(response) {
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:86]: Missing docblock

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 2]
```diff
@@ -11,16 +11,92 @@
...
+        // functions
+
+        function createServer() {
+            var emptyFields = _.some(
+              [vm.databaseForm.username, vm.databaseForm.password], function (x) {
+                  return x == '' || angular.isUndefined(x)
+              });
+
+            if (emptyFields) {
+                return
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: semicolon

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 3]
```diff
@@ -11,16 +11,92 @@
...
+        vm.completed = false;
+        vm.errors = [];
+
+        vm.createServer = createServer;
+        vm.connectServer = connectServer;
+
+        // functions
+
+        function createServer() {
+            var emptyFields = _.some(
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: This looks like error handling implementation. We normally have one single card for it, but on this particular page, we have 2:
https://trello.com/c/w4ZiBPUh/30-3-16-create-connect-crowbar-db-ui-story-2-page-and-navigation
https://trello.com/c/rJcjpRVv/39-2-15-create-connect-crowbar-db-error-handling

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 4]
```diff
@@ -11,16 +11,92 @@
...
+                        vm.completed = true;
+                    },
+                    // error
+                    function (errorResponse) {
+                        vm.errors = getErrorFromResponse(errorResponse);
+                    }
+                )
+                .finally(function () {
+                    vm.running = false;
+                })
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: semicolon

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 5]
```diff
@@ -11,16 +11,92 @@
...
+        }
+
+        function connectServer() {
+            var emptyFields = _.some(vm.databaseForm, function (x) {
+                return x == '' || angular.isUndefined(x)
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: `_.isUndefined(x) || _.isEmpty(x);`
Also, I'd change the variable name to `element` or `field` instead of `x`.

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js [hunk 6]
```diff
@@ -64,14 +94,124 @@ describe('Upgrade Flow - Create Connect Database Controller', function () {
...
+                    username: 'test',
+                    password: 'test'
+                };
+
+                controller.createServer();
+                $rootScope.$digest();
+            });
+
+            it('errors should be empty', function () {
+                expect(controller.errors).toEqual([]);
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js]: `expect(controller.errors).to.be.empty;`
http://chaijs.com/api/bdd/#method_empty

**Review by santiph — CHANGES_REQUESTED** [review]
A couple of questions and changes to be addressed

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js (hunk 6 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js]: weird. With this I get `undefined is not an object (evaluating 'expect(controller.errors).to.be')` but if I dump the object I get the proper `[]` ¯\_(ツ)_/¯

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 3 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: Indeed, this may be out of scope. I added it so it could easily be tested, but Im guessing that now that the ng-disabled directive is in place, its not needed anymore!

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 2 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: done

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 4 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: done

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 5 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: ah, thats great and much nicer to read! done!

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 3 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: oh wait, ng-disabled has nothing to do with it, this was to protect against empty fields when clicking the button and indeed, that should be dealt separately I guess? Althougth the form.isValid should enter in the scope for the card...button should be disabled until the fields are valid.

Thoughts? Im gonna leave it for the moment as its easy for testing, but is the form validation on the card scope?

**Comment by Itxaka** [comment]
[#14 Create/Connect Crowbar DB: Service Integration](https://trello.com/c/tslrmrBm/157-14-create-connect-crowbar-db-service-integration)

**Comment by Itxaka** [comment]
@santiph @skazi0 Ready for review

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 7]
```diff
@@ -11,16 +11,75 @@
...
+
+        vm.createServer = createServer;
+        vm.connectServer = connectServer;
+
+        // functions
+
+        function createServer() {
+            vm.errors = [];
+            vm.running = true;
+            vm.spinnerVisible = true;
```
Comment (skazi0) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: `spinnerVisible` should not be changed by the controller. It's a 'feedback' variable used by the spinner directive to indicate that the spinner is actually visible.

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 8]
```diff
@@ -11,16 +11,75 @@
...
+                )
+                .finally(function () {
+                    vm.running = false;
+                });
+        }
+
+        function connectServer() {
+            vm.errors = [];
+            vm.running = true;
+            vm.spinnerVisible = true;
```
Comment (skazi0) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: ditto

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js (hunk 6 above)
Comment (skazi0) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js]: @santiph this will not work as jasmine is overriding chai's `expect()`. To make this work you would need to locally switch to chai e.g. `var expect = window.chai.expect;` or call it directly via `chai.expect(...).to.be.empty;`.

File: assets/app/features/upgrade/i18n/en.json [hunk 9]
```diff
@@ -61,19 +61,19 @@
             },
             "database-configuration": {
                 "name": "",
-                "title": "Enter Credentials and/or Information to Create or Connect to an OpenStack PostgreSQL Database for the OpenStack Database Migration.",
+                "title": "Enter Credentials and/or Information to Create or Connect to an OpenStack PostgreSQL Server for the OpenStack Database Migration.",
                 "navigation-tabs": {
-                    "create-database": "Create New Database",
-                    "connect-database": "Connect to an Existing Database"
+                    "create-server": "Create New Server",
```
Comment (skazi0) [inline assets/app/features/upgrade/i18n/en.json]: Does it really create new _server_?

**Review by skazi0 — CHANGES_REQUESTED** [review]

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 7 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: Indeed!

File: assets/app/features/upgrade/i18n/en.json (hunk 9 above)
Comment (Itxaka) [reply assets/app/features/upgrade/i18n/en.json]: Yup!
It will run a chef-solo client to install and configure the postgres server on the administration server.

Still this will change in the future as @kwwii said that all text and so on needs to be refined, so I wouldnt give it much thougth rigth now.

Comment (Itxaka) [reply assets/app/features/upgrade/i18n/en.json]: well, not a _server_ per se but a _database server_ :+1:

Comment (santiph) [reply assets/app/features/upgrade/i18n/en.json]: We need to align to Ken's copys as specified in: https://github.com/crowbar/crowbar-ui/wiki/Upgrade67-Create-or-Connect-To-PostgreSQL-Database

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js (hunk 6 above)
Comment (santiph) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js]: Good point. @Itxaka I'd use `chai.expect` then

File: assets/app/features/upgrade/i18n/en.json (hunk 9 above)
Comment (Itxaka) [reply assets/app/features/upgrade/i18n/en.json]: Im guessing the refinement of this will come later, because there is a huge difference in what it says and in what it _does_.

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js (hunk 6 above)
Comment (Itxaka) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js]: done

**Review by skazi0 — APPROVED** [review]
lgtm

File: gulpfile.js [hunk 10]
```diff
@@ -41,6 +41,7 @@ gulp.task('default', function (callback) {
         'build',
         'rsync',
         'watch',
+        'tdd',
```
Comment (Itxaka) [inline gulpfile.js]: Mixing PR purpouse with extra stuff? Should be separated to a different PR (probably a slip?)

**Comment by Itxaka** [comment]
@santiph please re-review :D

File: gulpfile.js (hunk 10 above)
Comment (skazi0) [reply gulpfile.js]: yeah, sorry for that. reverted.

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 3 above)
Comment (santiph) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js]: Leave it as it is if that makes your life and development easier. :) Just know that this will be reviewed/adjusted when we work on the other cards.

File: assets/app/features/upgrade/i18n/en.json (hunk 9 above)
Comment (santiph) [reply assets/app/features/upgrade/i18n/en.json]: Until further specifications are required, we're supposed to implement what has been already defined in the comps

File: assets/app/data/crowbar/services/upgrade.factory.js [hunk 11]
```diff
@@ -81,7 +83,36 @@
             };
 
             return $http(requestOptions);
+        }
+
+        /**
+        * Install, configure and migrate a new database server on the administration server
```
Comment (santiph) [inline assets/app/data/crowbar/services/upgrade.factory.js]: Migrate a new database server? It doesn't makes much sense...

File: assets/app/data/crowbar/services/upgrade.factory.js [hunk 12]
```diff
@@ -81,7 +83,36 @@
...
+                url: '/api/upgrade/new',
+                headers: COMMON_API_V2_HEADERS,
+                data: data
+            };
+
+            return $http(requestOptions)
+        }
+
+        /**
+        * Configure and migrato to an existing database server
```
Comment (santiph) [inline assets/app/data/crowbar/services/upgrade.factory.js]: typo: `migrato` -> `migrate`

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 13]
```diff
@@ -11,20 +11,86 @@
...
-            tablePrefix: ''
+            host: '',
+            port: 5432
         };
-        // Set the current step as completed by default to enable the next button
-        // TODO(itxaka): remove this when we have the proper workflow for this step
-        upgradeStepsFactory.setCurrentStepCompleted();
 
+        /* eslint-disable max-len */
+        // NOTE: patterns based on https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/mixin/database.rb
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:26]: Thanks for this!

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js [hunk 14]
```diff
@@ -11,20 +11,86 @@
...
+            hostname: '^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$',
+            ipv4: '^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$',
+            username: '^[a-zA-Z0-9_]*$',
+            password: "^[a-zA-Z0-9_$&+,:;=?@#|'<>.^*()%!-]*$",
+        };
+        /* eslint-enable max-len */
+        vm.patterns.host = vm.patterns.hostname + '|' + vm.patterns.ipv4;
+
+        vm.running = false;
+        vm.spinnerVisible = false;
```
Comment (santiph) [inline assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:37]: @skazi0 Isn't this the attribute intended to be a notification only?

File: routes/api/upgrade/connect.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
+var express = require('express'),
+    router = express.Router();
+
+var responseOk = {
```
Comment (santiph) [inline routes/api/upgrade/connect.js]: Completely minor, specially for a mocked service like this: You can use the same `var` statement in line #1 like this:
```
var express = require('express'),
    router = express.Router(),
    responseOk = {
```

File: routes/api/upgrade/connect.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            }
+        },
+        'database_migration': {
+            'success': true
+        },
+        'schema_migration': {
+            'success': true
+        },
+        'crowbar_init': {
+            'success': true
```
Comment (santiph) [inline routes/api/upgrade/connect.js]: If the database setup step failed, shouldn't the database migration, schema migration and crowbar init tasks also not be successful?

**Review by santiph — CHANGES_REQUESTED** [review]
Partially reviewed, I'm still missing the view implementation (Will check it later today after some meetings).
Some changes requested.

**Comment by Itxaka** [comment]
rebased

File: assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js (hunk 14 above)
Comment (skazi0) [reply assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:37]: Not sure I understand your question. If you mean that this one doesn't need to be defined in the controller then yes, it can be done implicitly by angular.

File: assets/app/features/upgrade/templates/database-configuration-page.jade [hunk 17]
```diff
@@ -1,45 +1,83 @@
...
+    uib-tabset
+        uib-tab(index='0', disable='upgradeDatabaseVm.running || upgradeDatabaseVm.completed',
+        heading="{{'upgrade.steps.database-configuration.navigation-tabs.create-database' | translate}}")
+            form.database-form.new(name="newForm")
                 .row
                     .col-md-8
-                        .form-group
-                            label(translate='') upgrade.steps.database-configuration.database-form.username
-                            input.form-control(type='text', name='username', ng-required='false', ng-model='upgradeDatabaseVm.databaseForm.username')
+                        .form-group(ng-class="{'has-error': newForm.username.$invalid && newForm.username.$dirty}")
```
Comment (santiph) [inline assets/app/features/upgrade/templates/database-configuration-page.jade:13]: Is `has-error` actually needed, considering we already have `ng-invalid`, `ng-dirty` and `ng-touched`?
@see: https://docs.angularjs.org/guide/forms#using-css-classes

File: assets/app/features/upgrade/templates/database-configuration-page.jade [hunk 18]
```diff
@@ -1,45 +1,83 @@
...
+                                required='', ng-model='upgradeDatabaseVm.databaseForm.host')
                     .col-md-4
-                        .form-group
+                        .form-group(ng-class="{'has-error': existingForm.port.$invalid && existingForm.port.$dirty}")
                             .port
-                                label(translate='') upgrade.steps.database-configuration.database-form.port
-                                input.form-control(type='text', required='', ng-model='upgradeDatabaseVm.databaseForm.port')
+                                label.control-label(for='port', translate='') upgrade.steps.database-configuration.database-form.port
+                                input.form-control(type='number', name='port',
+                                min='1', max='65535',
```
Comment (santiph) [inline assets/app/features/upgrade/templates/database-configuration-page.jade:56]: Are `1` and `65365` allowed or excluded?

**Review by santiph — CHANGES_REQUESTED** [review]

File: assets/app/features/upgrade/templates/database-configuration-page.jade (hunk 18 above)
Comment (Itxaka) [reply assets/app/features/upgrade/templates/database-configuration-page.jade:56]: allowed

File: routes/api/upgrade/connect.js (hunk 16 above)
Comment (Itxaka) [reply routes/api/upgrade/connect.js]: I think I copied the response directly from the apiDocs so that may be the reason its like that? but indeed, all of them should be changed to failed

File: routes/api/upgrade/connect.js (hunk 15 above)
Comment (Itxaka) [reply routes/api/upgrade/connect.js]: woops, will fix

File: assets/app/data/crowbar/services/upgrade.factory.js (hunk 11 above)
Comment (Itxaka) [reply assets/app/data/crowbar/services/upgrade.factory.js]: lmao, indeed. Not sure what was I thinking when writing that, it makes no sense

File: routes/api/upgrade/connect.js (hunk 16 above)
Comment (skazi0) [reply routes/api/upgrade/connect.js]: The real question is: what will the backend return in such cases?

File: assets/app/features/upgrade/templates/database-configuration-page.jade (hunk 17 above)
Comment (skazi0) [reply assets/app/features/upgrade/templates/database-configuration-page.jade:13]: This is binding Angular's validation states with Bootstrap's classes: http://getbootstrap.com/css/#forms-control-validation

Comment (skazi0) [reply assets/app/features/upgrade/templates/database-configuration-page.jade:13]: As an alternative, we could consider using something like: https://www.npmjs.com/package/angular-bootstrap-form-validation. It handles `has-error` internally and in addition provides hints on the cause of validation failure.

Note that this one is not usable directly as it doesn't currently support `ngPattern` and has some old dependencies in `bower.json`. But maybe there are some newer alternatives with similar functionality. To be considered.

File: routes/api/upgrade/connect.js (hunk 16 above)
Comment (Itxaka) [reply routes/api/upgrade/connect.js]: Updated with the *actual* response provided in the apiDocs. Not sure where did I get this one from...

**Comment by Itxaka** [comment]
Updated to address the comments. Tested but its not working, button wont fire any event and no errors are shown

**Comment by Itxaka** [comment]
Disregard last comment, seems my cache had leftovers from olders patches. Cleaning and "gulping" again, everything works beautifully. Kudos to @skazi0 for the pattern matching, works wonderfully!

**Comment by santiph** [comment]
Looks good to me. I completely miss Bootstrap on the view you've implemented, thanks @skazi0 for remind me. `angular-bootstrap-form-validation` looks interesting, but I'd leave it as it is for now. We could analyse it and look for alternatives as part of the proper error handling cards we have.

**Review by santiph — APPROVED** [review]
